test/test_redirect_io.rb

```diff
@@ -0,0 +1,103 @@
+require_relative "helper"
+require_relative "helpers/integration"
+
+class TestRedirectIO < TestIntegration
+  parallelize_me!
+
+  def setup
+    super
+
+    @out_file = Tempfile.new 'puma-out'
+    @out_file_path = @out_file.path
+    @err_file = Tempfile.new 'puma-err'
+    @err_file_path = @err_file.path
+  end
+
+  def teardown
+    super
+
+    @out_file.close
```

Comment:
You can use `close!` and then you don't have to unlink manually.

Reply:
or `close(true)`

Reply:
`@out_file` is an instance of `File < IO`. There is no `close!` method and `IO#close` takes no arguments

https://ruby-doc.org/core-2.7.0/IO.html#method-i-close

Reply:
Oh tricky tricky, I didn't follow that through the test. Why replace a `Tempfile` with a `File`?

Reply:
After `log_rotate_output_files`, we reopen the file so we get a new handle to same path (otherwise the handle would still be pointing to the old file).

This was all to avoid having to create file handles repeatedly, but it was not really necessary and pretty confusing. I pushed a change to make it more clear.